### PR TITLE
More granular User API Key scopes, and plugin support

### PR DESCRIPTION
**Pull request by davidtaylorhq**
This PR includes three commits, which I intend to "rebase and merge", so that they remain separate. Please check each commit message for details. 

The overall aim here is to allow core/plugins to create a user api key, scoped to a specific route with specific parameters.

### CODE REVIEW

**Comment by danielwaterworth** [comment]
This is looking good, but it feels like the `RouteMatcher` is getting constructed at the wrong time. At the moment you're having to pass in `allowed_param_values` to the match method, but I think it would be clearer if the `RouteMatcher` had all of the data necessary to match against an arbitrary request. You could either achieve this by either allowing a `RouteMatcher` to be refined after construction, i.e. `RouteMatcher.new(...).with_allowed_param_values(...)` or not putting `RouteMatchers` into the constant.

**Comment by davidtaylorhq** [comment]
> it feels like the RouteMatcher is getting constructed at the wrong time. At the moment you're having to pass in allowed_param_values to the match method

All the parameters which `RouteMatcher` is initialized are constant. The "allowed param values" vary at runtime based on user-provided configuration. So they do need to be separated somehow, but we can certainly try and improve the ergonomics.

> You could achieve this by either allowing a RouteMatcher to be refined after construction, i.e. RouteMatcher.new(...).with_allowed_param_values(...)

I would be a little worried about making this thread-safe. `with_allowed_param_values` would have to return a fresh RouteMatcher to avoid polluting the unrefined RouteMatcher. Is that what you had in mind?

> not putting RouteMatchers into the constant.

Yeah I wasn't too sure about this... it feels a little weird. But one major advantage is that it means missing/extra initialisation parameters are detected at boot, rather than at runtime.

**Comment by danielwaterworth** [comment]
> I would be a little worried about making this thread-safe. with_allowed_param_values would have to return a fresh RouteMatcher to avoid polluting the unrefined RouteMatcher. Is that what you had in mind?

Yes, a fresh `RouteMatcher`.

I bring this up, because the name `RouteMatcher` implies to me that it has everything that it needs to match a route.

File: lib/auth/default_current_user_provider.rb [hunk 1]
```diff
@@ -336,7 +337,7 @@ def lookup_api_user(api_key_value, request)
       api_username = header_api_key? ? @env[HEADER_API_USERNAME] : request[API_USERNAME]
 
       params = @env['action_dispatch.request.parameters']
-      unless api_key.request_allowed?(request, params)
+      unless api_key.request_allowed?(@env)
```
Comment (romanrizzi) [inline lib/auth/default_current_user_provider.rb:340]: I think we don't need to assign the `params` variable anymore if we pass the env directly.

File: lib/plugin/instance.rb [hunk 2]
```diff
@@ -797,15 +797,33 @@ def add_api_key_scope(resource, action)
...
+    methods ||= method
+    actions ||= route
+    formats ||= format
+
+    DiscoursePluginRegistry.register_api_parameter_route(
+      RouteMatcher.new(
+        methods: methods,
+        actions: actions,
+        formats: formats
+      ), self)
```
Comment (romanrizzi) [inline lib/plugin/instance.rb:849]: Do we want to drop the old parameters at some point? If so, maybe we should a deprecation notice?

**Comment by davidtaylorhq** [comment]
Thanks @romanrizzi and @danielwaterworth - have made some improvements based on your comments.

**Comment by davidtaylorhq** [comment]
Fixed the failing specs, and updated the plugin API to automatically prefix scope names with the plugin name to prevent clashes (thanks for the suggestion @danielwaterworth)

**Pull request merged**